src/Streamly/Internal/Data/Producer/Type.hs

```diff
@@ -33,42 +35,66 @@ where
 #include "inline.hs"
 
 import Fusion.Plugin.Types (Fuse(..))
-import Streamly.Internal.Data.Stream.StreamD.Step (Step(..))
-import Prelude hiding (concat, map)
+import Prelude hiding (concat, map, const)
+
+-- $setup
+-- >>> :m
+-- >>> import qualified Streamly.Internal.Data.Producer as Producer
+-- >>> import qualified Streamly.Prelude as Stream
+
+------------------------------------------------------------------------------
+-- Step
+------------------------------------------------------------------------------
+
+{-# ANN type Step Fuse #-}
+data Step s a b = Yield b s | Skip s | Stop (Maybe a)
+
+instance Functor (Step s a) where
+    {-# INLINE fmap #-}
+    fmap f (Yield x s) = Yield (f x) s
+    fmap _ (Skip s) = Skip s
+    fmap _ (Stop s) = Stop s
 
 ------------------------------------------------------------------------------
 -- Type
 ------------------------------------------------------------------------------
 
+-- We need the return type of extract to be "Maybe a" so that we can represent
+-- a "used up" state for types that can not inherently represent that. For
+-- example, we may not be able to implement "const" correctly without it.
+--
+-- Or we can use the input type to be Maybe instead. Check duality with the
+-- type of "initial" in folds.
+--
 -- | A @Producer m a b@ is a generator of a stream of values of type @b@ from a
 -- seed of type 'a' in 'Monad' @m@.
 --
 -- /Internal/
 
 data Producer m a b =
     -- | @Producer step inject extract@
-    forall s. Producer (s -> m (Step s b)) (a -> m s) (s -> m a)
+    forall s. Producer (s -> m (Step s a b)) (a -> m s) (s -> m (Maybe a))
```

Comment:
This is probably a required change but I don't feel so right looking at it. Until we come up with a better abstraction this works.

Reply:
Is it possible to just modify the `Step` to represent this?